Tasks/JavaToolInstallerV0/javatoolinstaller.ts

```diff
@@ -81,35 +85,136 @@ async function getJava(versionSpec: string) {
     toolLib.prependPath(path.join(jdkDirectory, BIN_FOLDER));
 }
 
-function sleepFor(sleepDurationInMillisecondsSeconds): Promise<any> {
-    return new Promise((resolve, reeject) => {
-        setTimeout(resolve, sleepDurationInMillisecondsSeconds);
-    });
+/**
+ * Get file ending if it is supported. Otherwise throw an error.
+ * Find file ending, not extension. For example, there is supported .tar.gz file ending but the extension is .gz.
+ * @param file Path to a file.
+ * @returns string
+ */
+function getSupportedFileEnding(file: string): string {
+    const fileEnding: string = supportedFileEndings.find(ending => file.endsWith(ending)); 
+
+    if (fileEnding) {
+        return fileEnding;
+    } else {
+        throw new Error(taskLib.loc('UnsupportedFileExtension'));
+    }
 }
 
-function buildFilePath(localPathRoot: string, fileEnding: string, fileNameAndPath: string): string {
-    const fileName = fileNameAndPath.split(/[\\\/]/).pop();
-    const extractSource = path.join(localPathRoot, fileName);
+/**
+ * Install JDK.
+ * @param sourceFile Path to JDK file.
+ * @param fileExtension JDK file extension.
+ * @param archiveExtractLocation Path to folder to extract a JDK.
+ * @returns string
+ */
+async function installJDK(sourceFile: string, fileExtension: string, archiveExtractLocation: string, extendedJavaHome: string, versionSpec: string): Promise<string> {
+    let jdkDirectory;
+    if (fileExtension === '.dmg' && os.platform() === 'darwin') {
+        // Using set because 'includes' array method requires tsconfig option "lib": ["ES2017"]
+        const volumes: Set<string> = new Set(fs.readdirSync(VOLUMES_FOLDER));
+
+        await attach(sourceFile);
+    
+        const volumePath: string = getVolumePath(volumes);
+
+        let pkgPath: string = getPackagePath(volumePath);
+        try {
+            jdkDirectory = await installPkg(pkgPath, extendedJavaHome, versionSpec);
+        } catch (error) {
+            // In case of an error, there is still a need to detach the disk image
+            await detach(volumePath);
```

Comment:
Does it make sense to put it into finally block?

Reply:
Thank you! Changed it.